Utils/src/GenParticlesProducer.cc

```diff
@@ -99,7 +113,164 @@ void GenParticlesProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+
+    iEvent.put(std::move(genParticle_vec)); 
+    iEvent.put(std::move(PdgId_vec      ), "PdgId");
+    iEvent.put(std::move(Status_vec     ), "Status");
+    iEvent.put(std::move(ParentId_vec   ), "ParentId");
+    iEvent.put(std::move(Parent_vec     ), "Parent");
+    iEvent.put(std::move(TTFlag_vec     ), "TTFlag");
+}
+
+// Based on http://pdg.lbl.gov/2007/reviews/montecarlorpp.pdf
+enum particle_type
+{
+    down=1,
+    up=2,
+    strange=3,
+    charm=4,
+    bottom=5,
+    top=6,
+    electron=11,
+    electron_neutrino=12,
+    muon=13,
+    muon_neutrino=14,
+    tau=15,
+    tau_neutrino=16,
+    photon=22,
+    Z=23,
+    W=24,
+    Higgs=25
+};
+
+// Referenced:
+//   https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2017
+//   https://github.com/cms-sw/cmssw/blob/master/DataFormats/HepMCCandidate/interface/GenParticle.h
+//   https://github.com/cms-sw/cmssw/blob/master/DataFormats/Candidate/interface/CompositeRefCandidateT.h
+//   https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCandidateModules#ParticleTreeDrawer_Utility
+//   https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuidePATMCMatchingExercise
+const reco::GenParticle* GenParticlesProducer::findLast(const reco::GenParticle& particle) const {
+    const reco::GenParticle* last = &particle;
+    bool resetLast = false;
+    const reco::GenParticle *daughter;
+    if(debug) edm::LogInfo("TreeMaker") << "   Found a " << last->pdgId() << " (status=" << last->status() << ", ndaughters=" << last->numberOfDaughters() << "):";
+
+    while (true) {
+        resetLast=false;
+        for(unsigned int iDau=0; iDau<last->numberOfDaughters(); iDau++) {
+            daughter = static_cast<const reco::GenParticle *>(last->daughter(iDau));
+            if(daughter->pdgId()==last->pdgId()) {
+                last = daughter;
+                resetLast = true;
+                if(debug) edm::LogInfo("TreeMaker") << "      reset to " << last->pdgId() << "(status=" << last->status() << ")";
+            }
+        }
+        if(!resetLast) break;
+    }
+
+    /*
+    Empirical rules for particle status codes:
+    ==========================================
+    From a semi-leptonic TTJets sample:
+    -----------------------------------
+    1. tops start with status 22 and end with 62
+        any time there is one daughter it is simply radiating. when there are two daughters that should be the final top
+    2. b start with status 23 and end with 71/72 (ignore any gluon radiation)
+    3. W start with status 22, can immediately decay or go to status 51/52
+        immediately decays to status 23 which then decay to status 1/2 (leptonic) or 71 (hadronic)
+    */
+    if(((last->pdgId() >= down && last->pdgId() <= bottom) && (last->status()!=23 && last->status()!=51 && last->status()!=52 && last->status()!=71 && last->status()!=72 && last->status()!=73)) ||
```

Comment:
1. shouldn't all these be `abs(pdgId)`?
2. probably easier to put `pdgId` and `status` into temporary vars
3. for the list of quark status codes, make it a sorted vector and use `std::binary_search()`
4. are these status codes checked anywhere else, or is it just for this warning?

Reply:
1. Done
2. Done
3. Done
4. The status codes are only checked for that warning message. I thought about making it not store that depth if the codes weren't right, but since I'm still empirically finding new, acceptable code, that seemed prone to error.